# The Junction Usage Model (JUM)
- Please always refer to official release versions <https://github.com/qqwang-berkeley/JUM/releases> for stable performance JUM packages.
- JUM is a tool that offers a completely annotation-free approach to perform differential analysis of tissue-specific pre-mRNA alternative splicing patterns.

- JUM takes in RNA-seq data prepared from different tissues/cell samples and output alternative splicing events in six categories (cassette exon, mutually exclusive exons, alternative 5' splice site, alternative 3' splice site, intron retention, Composite patterns) that are significantly altered due to a specific treatment, or difference in tissue types, etc.

- JUM does not depend on any prior knowledge of the transcriptome annotation nor known libraries of AS events.

- The analysis functions in the current version are complete. I will continue to simplify the user-interface.

## Annouce the launch of JUM V2.0.2: featuring significant speed improvement (>20X), simplified user interface and updated user-friendly output formats.
* Stable performance packages for V2.0.2: <https://github.com/qqwang-berkeley/JUM/releases/tag/JUM_v2.0.2>
* Complete manual for V2.0.2: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-2.0.2-)>
* Legacy manual for V2.0 and V2.0.1: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-2.0-and-version-2.0.1)>

* Legacy manual for V1.3.12: <https://github.com/qqwang-berkeley/JUM/wiki/A-Hitchhiker's-guide-to-JUM-(version-1.3.12)>

A description of the JUM method is deposited to bioRxiv here (manuscript has been updated in May): https://www.biorxiv.org/content/early/2018/05/09/116863

Feel free to post any questions or comments on the Google group for JUM users: https://groups.google.com/forum/#!forum/jum-users
